### ISSUE TITLE
[Core] Actor not released when all references are released 

### ISSUE THREAD

**Issue by scv119** [ray-project/ray#24711]
### What happened + What you expected to happen

```
import ray

ray.init(num_cpus=2)

print(ray.available_resources())

@ray.remote(num_cpus=1)
class MyActor:
  def __init__(self, index):
    self.index = index

  def run(self):
    import time
    if self.index == 1:
      time.sleep(1)
      raise ValueError()
    else:
      time.sleep(1000)
      return 1

class ActorGroup:
  def run(self):
    actors = [MyActor.remote(i) for i in range(2)]
    futures = [a.run.remote() for a in actors]
    results = ray.get(futures)
    return results

ag = ActorGroup()
try:
    ag.run()
except Exception as e:
    print(e)

ag = ActorGroup()
ag.run()
```

In a two cpu cluster, this creates two actors, each of which takes 1 cpu, in an actor group,  releases the references to the actors; then creates another two actors. 

we expect the second round of actor creation to be successful because the first two actors went out of scope, but it actually hangs.

### Versions / Dependencies

nightly

### Reproduction script

see decription.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by scv119** [comment]
ah I guess the reason is because we don't force kill actor, when the actor is still running, even all the references went out of scope.

**Comment by amogkam** [comment]
I see, so this is expected behavior.

It does affect the experience on notebooks quite a bit. If I'm using Ray Datasets or Ray Train, or any library that orchestrates multiple actors, and one of them raises an error, I have to restart the entire notebook in order to free up the resources used by other actors.

**Comment by scv119** [comment]
> Actor processes will be terminated automatically when all copies of the actor handle have gone out of scope in Python, or if the original creator process dies.

https://docs.ray.io/en/latest/ray-core/actors/terminating-actors.html is the document, looks not reflect the actual semantic.

**Comment by scv119** [comment]
one concrete example @amogkam has is using libraries that manages actors (like dataset); user encountered some issues and tried to delete the dataset and creates a new one, but they might encounter this issue because some actor created by dataset is not killed in time.

**Comment by rkooo567** [comment]
Hmm what about in the destructor of ActorGroup, you call ray,kill of all actors?

**Comment by amogkam** [comment]
> Hmm what about in the destructor of ActorGroup, you call ray,kill of all actors?

I'm not sure if this will work all the time- the ActorGroup itself can still be in scope, even if the underlying actors are not. Basically you can change the last few lines of the code snippet above to this 
```
ag = ActorGroup()
try:
    ag.run()
except Exception as e:
    print(e)

# ag = ActorGroup()
ag.run()
```
and you will get the same problem, with the destructor of the ActorGroup never being called.

**Comment by amogkam** [comment]
So here's my thought process, let me know what you guys think

1. From the user perspective, this is pretty poor experience right now, especially on notebooks. To be able to understand what's going on, I need to know the following:
- How the libraries are using Ray actors
- How actor termination works (and even then the documentation is not fully correct as @scv119 pointed out).
So I think we should definitely fix this behavior for the user.

The question is, should this be fixed at the library layer or core layer? My opinion is that this might be too much of a burden for library developers to think about, and a much cleaner abstraction is just if actor handle goes out of scope, then it is terminated, without having to consider edge cases.

**Comment by amogkam** [comment]
> does provide a force kill mode (when out of scope) make sense? should we make that default behavior?

We should definitely source other opinions here, but I think making this the default behavior makes sense. This is what I thought was already the default behavior to begin with 😅

**Comment by amogkam** [comment]
Btw one short term solution could be to just raise a better error message. Not sure how difficult that is, but basically logging some messages if there is not enough resources in the cluster, and there are actors that are out of scope but are still executing tasks, on how the user can terminate these actors.

**Comment by scv119** [comment]
cc @jjyao ray 2.0 api clean up?

**Comment by rkooo567** [comment]
A couple of caveats I can think of when we make an ungrateful shutdown when it goes out of scope.

1. Semantically not consistent with other .remote APIs. E.g., When task or actor task ref goes out of scope the task is not canceled, but it will run. 
2. More prone to side effect because more tasks are "interrupted". If actor task has side effect, this could cause lots of unexpected consequences.

+1 we need more discussion & add this to 2.0 API clean up. We can also try to enable this feature and see how CI behaves (to find if there are unexpected edge cases). The implementation itself should be straightforward IIUC. 

Regarding error message, I am kind of against it. Feel like if we fix the doc, the behavior is just misuse of API, and I believe it can be fixed in the lib layer this way; 

```python3
class ActorGroup:
  def run(self):
    actors = [MyActor.remote(i) for i in range(2)]
    futures = [a.run.remote() for a in actors]
    try:
        results = ray.get(futures)
    except RayActorError as e:
        ray.kill(actors)
        raise e
    return results
```

**Comment by rkooo567** [comment]
Another factor to consider; Looks like Python thread has the same semantic as our actor (it doesn't forcefully cancel running tasks when the ref goes out of scope). I am also asking some other folks to learn the original motivation of this behavior now

**Comment by amogkam** [comment]
Got it, I see. Ok yeah we could resolve this at the library layer, but it would be great to document the behavior better.

I can make the change to Ray Train. @clarkzinzow @jianoaix are you guys OK with also implementing @rkooo567's code snippet above for Dataset's actor pool?

**Comment by rkooo567** [comment]
Btw, to be clear, I am not against changing the default semantics. I am just trying to figure out caveats for better decision. However, if it is urgent, doing it in lib layer sounds more reasonable since it could be pretty breaking changes (and it might take some time to make a decision from 2.0 API clean up efforts).

**Comment by amogkam** [comment]
Yep that sounds good! It would be great to understand these caveats some more, and have a broader discussion about this.

**Comment by jianoaix** [comment]
It looks to me out-of-scope destruction semantics is more clear in general, but this seems very C++-ish rather Python, which I think Sang pointed out the semantics consistency with other APIs we have.
For Dataset I think we can implement the fix. Just not sure non-ActorPool uses of actors that may still have this UX issue in Notebook-like context.

**Comment by matthewdeng** [comment]
> > Actor processes will be terminated automatically when all copies of the actor handle have gone out of scope in Python, or if the original creator process dies.
> 
> https://docs.ray.io/en/latest/ray-core/actors/terminating-actors.html is the document, looks not reflect the actual semantic.

Going back to this, was this behavior true at some point in time? Seems like the answer is yes (the docs do explicitly call out the behavior is implemented in Python but not Java/C++). 

Would someone be able to confirm if the behavior seen now is a regression or if it was an intentional change and the doc was never updated?
